Documentation/llilc-gc-transition.md

```diff
@@ -0,0 +1,96 @@
+2. It must have registered the last managed frame on its stack with the runtime
+   s.t. the GC may avoid scanning unmanaged frames.
+3. It must not reenter managed code if a GC is in progress.
+
+Threads that are running in preemptive mode need not be driven to a GC safe
+point before a collection, as invariants (1) and (3) guarantee that the
+managed code on such threads is already logically paused. Invariants (1) and
+(2) serve to satisfy the requirement that the GC must be able to scan the
+thread's stack for GC references.
+
+### Transitioning from managed and unmanaged code
+
+Transitioning from managed to unmanaged code boils down to switching the active
+thread from cooperative to preemptive mode while ensuring that the invariants
+associated with preemptive mode are obeyed. Abstractly, this looks something
+like the following, where `>> {Managed,Unmanaged} code` indicates the type of
+code currently running on the thread:
+
+```
+>> Managed code
+- Ensure all GC references in registers that might be updated by unmanaged code
+  have been spilled to locations that will not be updated by unmanaged code.
+  This satisfies the first preemptive mode invariant.
+- Register this frame as the last managed frame on this thread. This satisfies
+  the second preemptive mode invariant.
+- Flip this thread's GC mode to preemptive mode.
+
+>> Unmanaged code
+- Run unmanaged code. Usually this is just a call to an unmanaged function.
+- Once the unmanaged code has run, flip this thread's GC mode to cooperative
+  mode.
+
+>> Managed code
+- Check to see if a collection is in progress and if so, block this thread until
+  it completes. This satisfies the third preemptive mode invariant.
+```
+
+## Supporting GC transitions in LLILC and LLVM
+
+Ideally, LLILC would be able to emit all of the necessary transition code in
+LLVM IR. Unfortunately, it is not possible to do so while preserving the
+preemptive mode invariants:
+- The frame registration, mode transitions, and collection rendezvous must be
+  positioned immediately before and after the unmanaged code. LLVM IR has no
+  way of expressing this restriction.
+- Registering the managed frame requires knowing the address of the first
+  managed instruction following the unmanaged code. Again, LLVM IR has no
+  way of expressing this.
+
+As a result, the gc.statepoint intrinsic has been extended to support marking
+certain statepoints as transitions between code with differing GC models.
+These calls are then lowered differently by the backend, which is given an
+opportunity to lower the GC transition prologue and epilogue in a
+GC-model-specific fashion. In the case of CoreCLR, the backend is responsible
+for emitting the frame registration, mode transitions, and the collection
+rendezvous, while the frontend is resoponsible for allocating (and partially
+initializing) the frame registration data structure and marking calls to
+unamanged code as GC transitions.
```

Comment:
I also like the idea of adding an example IR. Irrelevant details in the dump can be masked if necessary.